### PR TITLE
Dropwizard testing module for JUnit 5

### PR DESCRIPTION
**Pull request by AnDyXX**
This PR is proposition of Dropwizard testing module with use of JUnit 5 (Issue [#1944](https://github.com/dropwizard/dropwizard/issues/1944)). Contains following:
- code of original test rules was changed to support executing them from JUnit 5 extension,
- JUnit 5 extension class that executes classes from point above,
- code of original tests modified to use JUnit 5.

Example of use:

```

@ExtendWith(DropwizardExtensionsSupport.class)
public class EpisodesResourceTest {

    public final ResourceExtension resources = ResourceExtension.builder()
            .addResource(new UsersResource(usersService))
            .build();
			
```

### CODE REVIEW

**Comment by nickbabcock** [comment]
Hey @AnDyXX, thanks for kicking this off! Much appreciated.

> I know that there is a lot of copied code from testing module that could be extracted to common testing library.

Is it possible to achieve backwards compatibility for the current junit4 users with a common core module? If not, we'd have to wait for something like a major version bump to introduce this change.

I'm guessing it's not possible to create the `dropwizard-testing-junit5` as a tacked on module to `dropwizard-testing` (eg. users poms would look like)

```xml
        <dependency>
            <groupId>io.dropwizard</groupId>
            <artifactId>dropwizard-testing</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>io.dropwizard</groupId>
            <artifactId>dropwizard-testing-junit5</artifactId>
            <scope>test</scope>
        </dependency>
```

As that might be the most preferable 🤔

**Comment by AnDyXX** [comment]
> Is it possible to achieve backwards compatibility for the current junit4 users with a common core module?

Let me try if this works.

**Comment by AnDyXX** [comment]
I changed code so JUnit4 and JUnit5 code is in the dropwizard-testing module. 
Changed junit 4 dependency from junit to junit-vintage-engine.
Common code extracted.

**Comment by joschi** [comment]
> Changed junit 4 dependency from junit to junit-vintage-engine.

Unfortunately that makes it incompatible with JUnit and would always pull in JUnit 5 and its vintage engine.

How about making the dependencies for JUnit 4 and 5 optional?
https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html

**Comment by AnDyXX** [comment]
Ok I made JUnit 5 optional and updated restored comments.

**Comment by joschi** [comment]
@AnDyXX Could you please rebase this PR on current `master`?

File: dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java [hunk 1]
```diff
@@ -34,7 +31,7 @@ long getId() {
     }
 
     String getDescription() {
-        return description;
+        return requireNonNull(description);
```
Comment (joschi) [inline dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java]: What is that for?

Throwing a `NullPointerException` in a *getter* is very bad practice.

File: dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java [hunk 2]
```diff
@@ -1,12 +1,9 @@
...
 
-import javax.persistence.Column;
-import javax.persistence.Entity;
-import javax.persistence.GeneratedValue;
-import javax.persistence.GenerationType;
-import javax.persistence.Id;
-import javax.persistence.Table;
-import javax.validation.constraints.NotNull;
+import javax.annotation.Nullable;
+import javax.persistence.*;
```
Comment (joschi) [inline dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java]: Please don't use wildcard imports but name all imports explicitly.

File: dropwizard-testing/src/test/java/io/dropwizard/testing/junit/TestEntity.java [hunk 3]
```diff
@@ -1,13 +1,7 @@
...
 
 import javax.annotation.Nullable;
-import javax.persistence.Column;
-import javax.persistence.Entity;
-import javax.persistence.GeneratedValue;
-import javax.persistence.GenerationType;
-import javax.persistence.Id;
-import javax.persistence.Table;
-import javax.validation.constraints.NotNull;
+import javax.persistence.*;
```
Comment (joschi) [inline dropwizard-testing/src/test/java/io/dropwizard/testing/junit/TestEntity.java]: Please don't use wildcard imports but name all imports explicitly.

File: dropwizard-testing/src/main/java/io/dropwizard/testing/common/DAOTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import io.dropwizard.logging.BootstrapLogging;
+import io.dropwizard.testing.junit.DAOTestRule;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.cfg.AvailableSettings;
+import org.hibernate.cfg.Configuration;
+import org.hibernate.context.internal.ManagedSessionContext;
+
+import java.util.*;
```
Comment (joschi) [inline dropwizard-testing/src/main/java/io/dropwizard/testing/common/DAOTest.java]: Please don't use wildcard imports but name all imports explicitly.

File: dropwizard-testing/src/main/java/io/dropwizard/testing/junit5/DropwizardExtensionsSupport.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
+package io.dropwizard.testing.junit5;
+
+import org.junit.jupiter.api.extension.*;
```
Comment (joschi) [inline dropwizard-testing/src/main/java/io/dropwizard/testing/junit5/DropwizardExtensionsSupport.java]: Please don't use wildcard imports but name all imports explicitly.

File: dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/DropwizardAppExtensionTest.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.core.MediaType;
+import java.io.PrintWriter;
+
+import static io.dropwizard.testing.ResourceHelpers.resourceFilePath;
+import static org.hamcrest.core.Is.is;
+import static org.hamcrest.MatcherAssert.*;
```
Comment (joschi) [inline dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/DropwizardAppExtensionTest.java]: Please don't use wildcard imports but name all imports explicitly.

File: dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/GzipDefaultVaryBehaviourTest.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import io.dropwizard.testing.junit.TestApplication;
+import io.dropwizard.testing.junit.TestConfiguration;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import javax.ws.rs.core.Response;
+import java.util.Collections;
+
+import static io.dropwizard.testing.ResourceHelpers.resourceFilePath;
+import static javax.ws.rs.core.HttpHeaders.*;
```
Comment (joschi) [inline dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/GzipDefaultVaryBehaviourTest.java]: Please don't use wildcard imports but name all imports explicitly.

File: dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/PersonResourceTest.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.core.GenericType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
```
Comment (joschi) [inline dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/PersonResourceTest.java]: Please don't use wildcard imports but name all imports explicitly.

File: dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
+package io.dropwizard.testing.junit5;
+
+import javax.annotation.Nullable;
+import javax.persistence.*;
```
Comment (joschi) [inline dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java]: Please don't use wildcard imports but name all imports explicitly.

File: dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java (hunk 1 above)
Comment (AnDyXX) [reply dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java]: Actually this one comes from this class: https://github.com/dropwizard/dropwizard/blob/master/dropwizard-testing/src/test/java/io/dropwizard/testing/junit/TestEntity.java

File: dropwizard-testing/src/test/java/io/dropwizard/testing/junit/TestEntity.java (hunk 3 above)
Comment (AnDyXX) [reply dropwizard-testing/src/test/java/io/dropwizard/testing/junit/TestEntity.java]: I addressed your comments, have to look at other failing issues.

**Comment by AnDyXX** [comment]
`codeclimate — 42 issues to fix` are mostly because of test for JUnit5 copied from junit. Any hints on this ones?

**Comment by arteam** [comment]
I've manually marked warnings in Codeclimate as false positives.

**Comment by AnDyXX** [comment]
PR code rebased.

**Comment by arteam** [comment]
Thank you for the contribution!

**Comment by krupeshf** [comment]
@AnDyXX / @arteam   

Do you have updated version of testing resources documentation?
https://www.dropwizard.io/1.3.9/docs/manual/testing.html#testing-resources  
Currently it is still using `JUnit4` and `ClassRule`  

Thank you :)

**Pull request merged**